tree/ntuple/v7/src/RPageStorageFile.cxx

```diff
@@ -0,0 +1,1414 @@
+   memcpy(&key, recordFooter, sizeof(key));
+   unsigned char *footer = new unsigned char[key.fObjLen];
+   fDecompressor(recordFooter + key.fKeyLen, key.fNbytes - key.fKeyLen, key.fObjLen, footer);
+   delete[] recordFooter;
+
+   RNTupleDescriptorBuilder descBuilder;
+   descBuilder.SetFromHeader(header);
+   descBuilder.AddClustersFromFooter(footer);
+   delete[] header;
+   delete[] footer;
+
+   return descBuilder.MoveDescriptor();
+}
+
+
+ROOT::Experimental::RNTupleDescriptor ROOT::Experimental::Detail::RPageSourceFile::AttachRaw()
+{
+   RRawFileHeader fileHeader;
+   Read(&fileHeader, sizeof(fileHeader), 0);
+   RTFString ntupleName;
+   auto offset = sizeof(fileHeader);
+   Read(&ntupleName, 1, offset);
+   offset += ntupleName.GetSize();
+   RTFNTuple ntuple;
+   Read(&ntuple, sizeof(ntuple), offset);
+
+   auto recordHeader = new unsigned char[ntuple.fNBytesHeader];
+   auto header = new unsigned char[ntuple.fLenHeader];
+   Read(recordHeader, ntuple.fNBytesHeader, ntuple.fSeekHeader);
+   fDecompressor(recordHeader, ntuple.fNBytesHeader, ntuple.fLenHeader, header);
+   delete[] recordHeader;
+
+   auto recordFooter = new unsigned char[ntuple.fNBytesFooter];
+   auto footer = new unsigned char[ntuple.fLenFooter];
+   Read(recordFooter, ntuple.fNBytesFooter, ntuple.fSeekFooter);
+   fDecompressor(recordFooter, ntuple.fNBytesFooter, ntuple.fLenFooter, footer);
+   delete[] recordFooter;
+
+   RNTupleDescriptorBuilder descBuilder;
+   descBuilder.SetFromHeader(header);
+   descBuilder.AddClustersFromFooter(footer);
+   delete[] header;
+   delete[] footer;
+   return descBuilder.MoveDescriptor();
+}
+
+
+ROOT::Experimental::RNTupleDescriptor ROOT::Experimental::Detail::RPageSourceFile::DoAttach()
+{
+   char ident[4];
+   Read(ident, 4, 0);
+   if (std::string(ident, 4) == "root")
+      return AttachTFile();
+
+   return AttachRaw();
+}
+
+
+ROOT::Experimental::Detail::RPage ROOT::Experimental::Detail::RPageSourceFile::PopulatePageFromCluster(
+   ColumnHandle_t columnHandle, const RClusterDescriptor &clusterDescriptor, ClusterSize_t::ValueType clusterIndex)
+{
+   auto columnId = columnHandle.fId;
+   auto clusterId = clusterDescriptor.GetId();
+   const auto &pageRange = clusterDescriptor.GetPageRange(columnId);
+
+   // TODO(jblomer): binary search
+   RClusterDescriptor::RPageRange::RPageInfo pageInfo;
+   decltype(clusterIndex) firstInPage = 0;
+   for (const auto &pi : pageRange.fPageInfos) {
+      if (firstInPage + pi.fNElements > clusterIndex) {
+         pageInfo = pi;
+         break;
+      }
+      firstInPage += pi.fNElements;
+   }
+   R__ASSERT(firstInPage <= clusterIndex);
+   R__ASSERT((firstInPage + pageInfo.fNElements) > clusterIndex);
+
+   auto element = columnHandle.fColumn->GetElement();
+   auto elementSize = element->GetSize();
+
+   auto pageSize = pageInfo.fLocator.fBytesOnStorage;
+   void *pageBuffer = malloc(std::max(pageSize, static_cast<std::uint32_t>(elementSize * pageInfo.fNElements)));
```

Comment:
Why malloc rather than operator new [] ? (either way, it might be helpful to add the answer as comment in the code).